src/message/inbound.rs

```diff
@@ -130,8 +57,101 @@ impl<'a> Message<'a> for InboundMessage {
     }
 }
 
-impl InboundMessage {
-    pub fn get_receive_timestamp(&self) -> Result<Option<SystemTime>> {
+pub struct FlowInboundMessage {
+    _msg_ptr: ffi::solClient_opaqueMsg_pt,
+    _flow_ptr: ffi::solClient_opaqueFlow_pt,
+}
+
+impl InboundMessageTrait<'_> for FlowInboundMessage {}
+
+impl fmt::Debug for FlowInboundMessage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut f = f.debug_struct("FlowInboundMessage");
+        debug_inbound_message_fields(self, &mut f);
+        f.finish()
+    }
+}
+
+unsafe impl Send for FlowInboundMessage {}
+
+impl Drop for FlowInboundMessage {
+    fn drop(&mut self) {
+        let rc = unsafe { ffi::solClient_msg_free(&mut self._msg_ptr) };
+
+        let rc = SolClientReturnCode::from_raw(rc);
+        if !rc.is_ok() {
+            warn!("warning: message was not dropped properly");
+        }
+    }
+}
+
+impl From<(ffi::solClient_opaqueMsg_pt, ffi::solClient_opaqueFlow_pt)> for FlowInboundMessage {
+    /// .
+    ///
+    /// # Safety
+    ///
+    /// From a valid owned pointer.
+    /// No other alias should exist for this pointer
+    /// FlowInboundMessage will try to free the ptr when it is destroyed
+    ///
+    /// .
+    fn from(
+        (_msg_ptr, _flow_ptr): (ffi::solClient_opaqueMsg_pt, ffi::solClient_opaqueFlow_pt),
+    ) -> Self {
+        Self {
+            _msg_ptr,
+            _flow_ptr,
+        }
+    }
+}
+
+impl<'a> Message<'a> for FlowInboundMessage {
+    unsafe fn get_raw_message_ptr(&self) -> ffi::solClient_opaqueMsg_pt {
+        self._msg_ptr
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum FlowInboundMessageAckError {
+    #[error("Invalid message: subcode {0}")]
+    InvalidMessage(SolClientSubCode),
+    #[error("Message not found")]
+    MessageNotFound,
+    #[error("Ack failed: subcode {0}")]
+    AckFailed(SolClientSubCode),
+}
+
+impl FlowInboundMessage {
+    pub fn try_ack(&self) -> std::result::Result<(), FlowInboundMessageAckError> {
+        let mut message_id: solClient_msgId_t = 0;
+        let get_message_id_return_code = unsafe {
+            let get_message_id_return_code_raw =
+                ffi::solClient_msg_getMsgId(self._msg_ptr, &mut message_id);
+            SolClientReturnCode::from_raw(get_message_id_return_code_raw)
+        };
+        if let SolClientReturnCode::NotFound = get_message_id_return_code {
+            return Err(FlowInboundMessageAckError::MessageNotFound);
+        }
+        if !get_message_id_return_code.is_ok() {
+            return Err(FlowInboundMessageAckError::InvalidMessage(
+                get_last_error_info(),
+            ));
+        }
+
+        let send_ack_return_code = unsafe {
+            let send_ack_return_code_raw = ffi::solClient_flow_sendAck(self._flow_ptr, message_id);
```

Comment:
If each flow message stores a ptr to the flow object and the flow object gets dealloc, this method will cause a use after free.

Reply:
It's hard to deal with raw pointers

What are your thoughts on reverting `FlowInboundMessage` that has `flow_ptr`, and instead provide `Ack` function to on_message callback?

```rust
// in src/flow/callback
extern "C" fn static_on_message<'s, F, Ack>(
    opaque_flow_p: ffi::solClient_opaqueFlow_pt,   // non-null
    msg_p: ffi::solClient_opaqueMsg_pt,            // non-null
    raw_user_closure: *mut ::std::os::raw::c_void, // can be null
) -> ffi::solClient_rxMsgCallback_returnCode_t
where
    F: FnMut(InboundMessage, &Ack) + Send + 's,
    Ack: FnMut() -> Result<(), AckError>,
{
    ...
    let ack_fn = || { ... };
    user_closure(message, &ack_fn);
    ...
}
```

Reply:
Sorry about not getting to this. I was considering that it might be easier to reason about if the ack function is part of the flow struct rather than flow message containing a reference for the flow struct. The ack function could either accept a message id or message itself and it would be up to the caller to make sure they are able to keep flow struct long enough to send the acknowledgement.